### ISSUE TITLE
Add good pratices to train bert model in documentation

### ISSUE THREAD

**Issue by emilienschultz** [emilienschultz/activetigger#372]
Résumé de la réunion avec @jboelaert @julesbrion et Annina

Améliorer la compréhension des différents types de modèles car cela perd un peu. Modèle rapide ; modèle bert ; …
Mentionner le fait que la visualisation permet de désambiguiser certains textes
Mentionner que l’entropie (à la fois pour le modèle rapide et la prédiction exportable) permet de donner une mesure de l’incertitude.
Mentionner qu’il vaut mieux éviter de modifier les annotations sur la base des prédictions : la règle principale doit toujours être la cohérence avec le codebook et non pas ce que prédit le modèle (à la marge, cela permet d’identifier des problèmes dans le codage).
Une bonne pratique pour utiliser les modèles rapides : 1/ commencer avec les features sbert puis 2/ utiliser les prédictions du dernière modèle BERT dès que le modèle commence à avoir des résultats correctes.
Bonnes pratiques de gestion de projet : penser à supprimer les projets inutilisés ; ne pas garder tous les modèles car ils prennent de la place

La démarche générale pour entrainer un modèle : 
0/ avoir quelques centaines d’annotations (environ 50 par classes)
1/ partir des paramètres par défaut qui devraient être bien choisis
2/ regarder la courbe de la loss avec deux principales règles : (1) il faut que la loss décroissent ; (2) dans l’idéal les deux courbes devraient avoir des allures similaires et (3) il faut minimiser la loss
Pour modifier l’allure, il est possible d’utiliser les paramètres

Dans un premier temps
-	Si la courbe ne décroit pas et reste stable, cela signifie que l’apprentissage n’a pas lieu, et il faut augmenter le learning rate
-	Si la courbe croit au lieu de décroitre, le modèle apprend trop vite (et « désapprend » de son préentrainement), il faut diminuer le learning rate
-	Si l’évolution est chaotique, cela signifie que le modèle dépend trop du bruit, il faut plutôt augmenter le learning rate
-	Si la loss est plus basse à la fin que la loss d’éval, c’est que le modèle a surappris 
-	Si la loss reste plate, c’est que l’apprentissage n’avance plus, il est possible de réduire le nombre d’epoch
Dans un second temps, on utilise le learning rate pour minimiser la loss
On peut jouer sur le batchsize et le grad acc qui permettent de régulariser le comportement, les deux sont liés (la multiplication des deux donnent le batch réel), donc sont utilisés ensemble : final batch size affiche la multiplication des deux

+ Dans la documentation, mettre un exemple de courbe « correctes » et un exemple de courbes « problématiques » 
+ préciser ce que veut dire best model : le modèle retenu est celui avec la meilleure éval
+ dans la doc mettre des ressources 
-	https://developers.google.com/machine-learning/crash-course/overfitting/interpreting-loss-curves?hl=fr
-	https://wandb.ai/mostafaibrahim17/ml-articles/reports/A-Deep-Dive-Into-Learning-Curves-in-Machine-Learning--Vmlldzo0NjA1ODY0

**Comment by ninclay** [comment]
Hello,

Proposing some text for a new section on best practices in model training for the QuickStart based on the above notes. Does this seem accurate/does this make sense? What to add/change/remove?

**Suggested Workflow for Model Training**

Once you have annotated a sufficient number of examples (ideally a few hundred in total, at least around 50 per class/category) you can begin training classification models to assist and scale your annotation work. ActiveTigger supports a progressive workflow for model training. 

**Training Quick Models to Evaluate Intermediary Performance**
You can train one or more "quick", intermediary models to see how well your model is performing based on your existing annotations. These can provide preliminary insights on baseline performance. This can help you in early stages of annotation, identifying cases where a model may struggle on ambiguities or other inconsistencies in your coding scheme/data. 

1. Define your features under Prepare tab (we recommend SBERT). This is a pre-trained model that converts your text into a numerical representation (embeddings). This allows you to use the **Visualization** feature to see how this preliminary model groups texts. A clear visual separation indicates that the model is capturing how your categories should be distinguished.
2. You can follow the steps in **Fine-tune your BERT** classifier to train your first models on your actual annotations, even with a more limited annotation set. See guide on validation metrics for ways to assess performance.
3. Once you start to have decent results in your first trained models, you can use the predictions of your latest model in visualizations, prediction export, and further annotation (active learning). In particular, the loss curve is a useful evaluation tool (see below). The entropy score (provided alongside predictions) can also be a useful metric for evaluation at this stage. High entropy suggests low confidence, helping you pinpoint ambiguous cases or inconsistencies in your annotations.

Box:
It can be tempting to adjust your annotations according to initial performance metrics (for example, a lower F1 score in a certain category may incite you to tag more examples as positives for this category). **Ensure that your annotations are first and foremost consistent with your codebook.** Mixing annotation strategies in the middle of the same scheme can cause issues with training models due to inconsistencies. If you are unhappy with model results, consider adjusting your codebook - maybe your categories are too vaguely defined? You can start a new coding scheme to test various annotation strategies.

**Monitoring the Training Process**
During model training, ActiveTigger displays loss curves for both training and evaluation datasets. These curves help you assess the model's learning progress:

Decreasing Loss: Indicates effective learning. If the loss remains flat, consider increasing the learning rate.
Increasing Loss: Suggests the learning rate may be too high, try lowering it.
Chaotic Loss Patterns: May indicate the model is learning from noise; adjust the learning rate accordingly.
Overfitting: If the evaluation loss is significantly higher than the training loss, the model may be overfitting (= the model "memorizes" rather than "learns" and will perform poorly on unseen data).
Stalled Learning: If both curves are flat, consider reducing the number of epochs.

(screenshot examples of curves with descriptions)

You can also adjust the batch size (= how many examples the model looks at before updating itself) and gradient accumulation (= simulating large-batch training). The product of these two settings gives the effective batch size, which is displayed in the interface.

These evaluation tools can help you define the parameters for your final model to be applied on the whole dataset.

(...)
Best Practices in Storage
Please make sure to regularly delete unused models and archive projects no longer in active use, as these will take up unnecessary space.

**Comment by emilienschultz** [comment]
@jboelaert if you have some time, could you have a look on this text for the Quickstart ?

**Comment by emilienschultz** [comment]
Hi @ninclay it seems a nice discussion text point what you wrote. You can directly add it in the .md files, so we can make suggestion on it (I can assist you for that if you need).

Just 2 sentences to add before explaining the different types of curves : 1/ that the goal is to minimize loss, i.e. the difference between the data we have and the data we predict ; 2/ the difference between eval/model : eval is computed on a separate sample to have an a mesrue without interference of the training data (not sure if I phrase it well; but it is a starting point).

**Comment by emilienschultz** [comment]
@ninclay if I am not mistaken, you merged all this information in the documentation. Thanks !! So I can close this issue. Tell me if it is not the case.
